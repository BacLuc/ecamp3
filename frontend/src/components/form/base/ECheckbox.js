import { VCheckbox } from 'vuetify/lib'

export default {
  name: 'ECheckbox',
  extends: VCheckbox,
  props: {
    hideDetails: {
      type: String,
      default: 'auto'
    },
    noMargin: {
      type: Boolean,
      default: false,
      required: false
    }
  },
  computed: {
    classes () {
      return {
        ...VCheckbox.options.computed.classes.call(this),
        'my-4': !this.noMargin
      }
    }
  }
}
