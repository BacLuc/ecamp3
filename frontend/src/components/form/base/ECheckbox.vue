<template>
  <ValidationProvider v-slot="{ errors }" :name="name"
                      :vid="veeId"
                      :rules="veeRules">
    <v-checkbox v-bind="$attrs"
                :hide-details="hideDetails"
                :error-messages="errors"
                :label="label || name"
                :class="'my-' + my"
                v-on="$listeners" />
  </ValidationProvider>
</template>

<script>
import { ValidationProvider } from 'vee-validate'
import { formComponentPropsMixin } from '@/mixins/formComponentPropsMixin'

export default {
  name: 'ECheckbox',
  components: { ValidationProvider },
  mixins: [formComponentPropsMixin]
}
</script>
