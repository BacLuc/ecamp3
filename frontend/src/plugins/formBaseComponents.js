// Global registration of all Vue components in folder components/form/base

// import upperFirst from 'lodash/upperFirst'
// import camelCase from 'lodash/camelCase'

import ECheckbox from '@/components/form/base/ECheckbox.vue'
import EColorPicker from '@/components/form/base/EColorPicker.vue'
import EDatePicker from '@/components/form/base/EDatePicker.vue'
import ERichtext from '@/components/form/base/ERichtext.vue'
import ESelect from '@/components/form/base/ESelect.vue'
import ESwitch from '@/components/form/base/ESwitch.vue'
import ETextarea from '@/components/form/base/ETextarea.vue'
import ETextField from '@/components/form/base/ETextField.vue'
import ETimePicker from '@/components/form/base/ETimePicker.vue'

/*
const requireComponent = require.context(
  // The relative path of the components folder
  '../components/form/base',
  // Whether or not to look in subfolders
  false,
  // The regular expression used to match base component filenames
  /[A-Z]\w+\.(vue|js)$/
) */

class FormBaseComponentsPlugin {
  install (Vue, options) {
    Vue.component('ECheckbox', ECheckbox)
    Vue.component('EColorPicker', EColorPicker)
    Vue.component('EDatePicker', EDatePicker)
    Vue.component('ERichtext', ERichtext)
    Vue.component('ESelect', ESelect)
    Vue.component('ESwitch', ESwitch)
    Vue.component('ETextarea', ETextarea)
    Vue.component('ETextField', ETextField)
    Vue.component('ETimePicker', ETimePicker)

    /*
    requireComponent.keys().forEach(fileName => {
      // Get component config
      const componentConfig = requireComponent(fileName)

      // Get PascalCase name of component
      const componentName = upperFirst(
        camelCase(
          // Gets the file name regardless of folder depth
          fileName
            .split('/')
            .pop()
            .replace(/\.\w+$/, '')
        )
      )

      // Register component globally
      Vue.component(
        componentName,
        // Look for the component options on `.default`, which will
        // exist if the component was exported with `export default`,
        // otherwise fall back to module's root.
        componentConfig.default || componentConfig
      )
    }) */
  }
}

export default new FormBaseComponentsPlugin()
