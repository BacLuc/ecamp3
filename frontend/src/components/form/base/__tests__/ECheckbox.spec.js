import Vue from 'vue'
import Vuetify from 'vuetify'

import formBaseComponents from '@/plugins/formBaseComponents'

import { mount as mountComponent } from '@vue/test-utils'
import ECheckbox from '../ECheckbox'

Vue.use(Vuetify)
Vue.use(formBaseComponents)

describe('An ECheckbox', () => {
  let vuetify

  const mount = (options) => mountComponent(ECheckbox, { vuetify, ...options })
  beforeEach(() => {
    vuetify = new Vuetify()
  })
  test('looks like a checkbox', async () => {
    const wrapper = mount()
    await wrapper.setProps({ vModel: false })
    expect(wrapper.element).toMatchSnapshot('unchecked')

    await wrapper.setProps({ vModel: true })
    expect(wrapper.element).toMatchSnapshot('checked')
  })

  test('is checked when initialized checked', () => {
    const wrapper = mount({
      propsData: {
        vModel: true
      }
    })
    expect(wrapper.find('input[type=checkbox]').element.getAttribute('vmodel')).toBe('true')
  })

  test('updates checkbox when vModel changes', async () => {
    const wrapper = mount()
    expect(wrapper.find('input[type=checkbox]').element.getAttribute('vmodel')).toBeNull()

    await wrapper.setProps({ vModel: true })
    expect(wrapper.find('input[type=checkbox]').element.getAttribute('vmodel')).toBe('true')

    await wrapper.setProps({ vModel: false })
    expect(wrapper.find('input[type=checkbox]').element.getAttribute('vmodel')).toBeNull()
  })

  test('updates vModel when user clicks on checkbox', async () => {
    const wrapper = mount()
    const input = wrapper.find('input')

    const change = jest.fn()
    wrapper.vm.$on('input', (event) => change(event))

    await input.trigger('click')
    expect(change).toBeCalledTimes(1)
    expect(change).toBeCalledWith(true)

    jest.resetAllMocks()
    await input.trigger('click')
    expect(change).toBeCalledTimes(1)
    expect(change).toBeCalledWith(false)
  })
})
