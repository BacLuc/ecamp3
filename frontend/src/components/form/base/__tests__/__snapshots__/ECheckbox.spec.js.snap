// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`An ECheckbox looks like a checkbox: checked 1`] = `
<span>
  <div
    class="v-input v-input--hide-details theme--light v-input--selection-controls v-input--checkbox"
  >
    <div
      class="v-input__control"
    >
      <div
        class="v-input__slot"
      >
        <div
          class="v-input--selection-controls__input"
        >
          <i
            aria-hidden="true"
            class="v-icon notranslate mdi mdi-checkbox-blank-outline theme--light"
          />
          <input
            aria-checked="false"
            id="input-8"
            role="checkbox"
            type="checkbox"
            value=""
            vmodel="true"
          />
          <div
            class="v-input--selection-controls__ripple"
          />
        </div>
      </div>
    </div>
  </div>
</span>
`;

exports[`An ECheckbox looks like a checkbox: unchecked 1`] = `
<span>
  <div
    class="v-input v-input--hide-details theme--light v-input--selection-controls v-input--checkbox"
  >
    <div
      class="v-input__control"
    >
      <div
        class="v-input__slot"
      >
        <div
          class="v-input--selection-controls__input"
        >
          <i
            aria-hidden="true"
            class="v-icon notranslate mdi mdi-checkbox-blank-outline theme--light"
          />
          <input
            aria-checked="false"
            id="input-8"
            role="checkbox"
            type="checkbox"
            value=""
          />
          <div
            class="v-input--selection-controls__ripple"
          />
        </div>
      </div>
    </div>
  </div>
</span>
`;
